IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/ElementsHandler.cs

```diff
@@ -98,21 +111,52 @@ public bool ContainsText(string text)
         /// </summary>
         /// <param name="predicate">The criteria to attempt to match on.</param>
         /// <returns></returns>
-        public IWebElement GetSingleExistingElement(Func<IWebElement, bool> predicate)
+        public IWebElement GetSingleWebElement(Func<IWebElement, bool> predicate)
+        {
+            IList<IWebElement> elems = GetElements(predicate);
+
+            if (elems.Count != 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(predicate), "The provided predicate did not match exactly one result.");
+            }
+
+            return elems[0];
+        }
+
+        /// <summary>
+        /// Gets all elements found by <see cref="Locator"/>, matching a given predicate.
+        /// </summary>
+        /// <param name="predicate">The function used to filter to one or more IWebElements</param>
+        /// <returns>A list of found IWebElements</returns>
+        /// <exception cref="NoSuchElementException"></exception>
+        /// <exception cref="ArgumentOutOfRangeException"></exception>
+        public IList<IWebElement> GetAllWebElements(Func<IWebElement, bool> predicate)
+        {
+            return GetElements(predicate);
+
+        }
+
+        public IList<IWebElement> GetElements(Func<IWebElement, bool> predicate)
         {
             IWait<IWebDriver> wait = Wait;
             wait.IgnoreExceptionTypes(typeof(NoSuchElementException));
-            return wait.Until(d =>
+
+#pragma warning disable CS8603 // Possible null reference return. Needed for Selenium retry.
+            return wait.Until(_ =>
             {
-                var foundElems = d.FindElements(Locator);
-                if (foundElems is null || foundElems.Count == 0) throw new NoSuchElementException($"No element found matching pattern: {Locator}");
-                var foundElem = foundElems.Where(predicate).ToList();
-                if (foundElem.Count != 1)
+                IReadOnlyCollection<IWebElement> foundElems = SearchContext.FindElements(Locator);
+                if (foundElems is null || foundElems.Count == 0) throw new NoSuchElementException($"No elements found matching pattern: {Locator}");
+                IList<IWebElement> elements = foundElems.Where(predicate).ToList();
```

Comment:
Should the predicated be applied before the if?

Reply:
Maybe a naming problem? The goal of the method is:

For a given selenium locator, for the elements Selenium returns, filter down those returned elements to some more precise list.

I'm doing the if as a final check to essentially make sure at least one element is left after doing the subsequent filtering. Does that make sense, or is something still not adding up?

Reply:
Oh wait, I think I see what you're saying. I could potentially turn this into a single elements.Any() (one of your other comments confirmed Selenium's FindElements shouldn't ever return null,) check instead of two different ifs and get the same functionality. I'll think about that while I sleep.

Reply:
Moved the predicate to be applied inline with the FindElements, simplified the logic to a single Any() call.